### PR TITLE
Initial code

### PR DESCRIPTION
**Pull request by karmanyaahm**
To make this ready to publish to pub.dev the git dependency will need to be removed, and for that, webcrypto will need to publish another version based on their latest master branch.

### CODE REVIEW

**Comment by karmanyaahm** [comment]
@emersion I can't figure out how to request a review from you in the GitHub review section, but can you check this out whenever you're free?

File: lib/webpush_encryption.dart [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+    return WebPushKeys(
+            pubKeyBytes: base64Decode(split[0]),
+            authKeyBytes: base64Decode(split[1]),
+            privKeyBytes: base64Decode(split[2]))
+        ._validate();
+  }
+
+  /// Generate a new, random keyset.
+  static Future<WebPushKeys> random() async {
```
Comment (p1gp1g) [inline lib/webpush_encryption.dart]:
```suggestion
  static Future<WebPushKeys> newKeyPair() async {
```

File: lib/webpush_encryption.dart [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+// getters
+  /// Get key in webcrypto format
+  Future<EcdhPublicKey> get pubKey =>
+      EcdhPublicKey.importRawKey(_pubKey, curve);
+
+  /// aka `p256dh`. Get the base64Url encoded public key to send to application server.
+  ///
+  ///This should be paired with [authWeb].
+  String get pubKeyWeb => base64UrlEncode(_pubKey);
```
Comment (p1gp1g) [inline lib/webpush_encryption.dart]:
```suggestion
  String get p256dh => base64UrlEncode(_pubKey);
```

File: lib/webpush_encryption.dart [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+  /// aka `auth`. Get the base64Url encoded authentication key to send to application server.
+  ///
+  /// This should be paired with [pubKeyWeb].
+  String get authWeb => base64UrlEncode(_authKey);
```
Comment (p1gp1g) [inline lib/webpush_encryption.dart]:
```suggestion
  String get auth => base64UrlEncode(_authKey);
```

File: lib/webpush_encryption.dart [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    }
+    return WebPushKeys(
+            pubKeyBytes: pub!,
+            authKeyBytes: Uint8List.fromList(auth!),
+            privKeyBytes: priv!)
+        ._validate();
+  }
+
+  /// Deserializes the output of [toBase64]. Used for retrieval of keys from storage.
+  static Future<WebPushKeys> fromBase64(String base64str) {
```
Comment (p1gp1g) [inline lib/webpush_encryption.dart]:
```suggestion
  static Future<WebPushKeys> deserialize(String base64str) {
```

File: lib/webpush_encryption.dart [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+// export
+  Map<String, List<int>> get allKeysRaw => {
+        'p256dh': _pubKey,
+        'auth': _authKey,
+        'priv': _privKey,
+      };
+
+  /// Serializes public *and private* keys for storage. Deserialize with [WebPushKeys.fromBase64()]
+  String get toBase64 =>
```
Comment (p1gp1g) [inline lib/webpush_encryption.dart]:
```suggestion
  String get serialize =>
```

File: lib/webpush_encryption.dart [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  ///
+  /// This should be paired with [pubKeyWeb].
+  String get authWeb => base64UrlEncode(_authKey);
+
+  ///Get key in webcrypto format
+  Future<EcdhPrivateKey> get privKey =>
+      EcdhPrivateKey.importPkcs8Key(_privKey, curve);
+
+// export
+  Map<String, List<int>> get allKeysRaw => {
```
Comment (p1gp1g) [inline lib/webpush_encryption.dart]:
```suggestion
  Map<String, List<int>> get rawKeys => {
```

**Review by p1gp1g — CHANGES_REQUESTED** [review]
It looks great. I just suggest to change some functions name

**Pull request merged**